### PR TITLE
ci: add Qwen 3.6 Plus security review via OpenRouter

### PR DESCRIPTION
**Pull request by 2tbmz9y2xt-lang**
## Summary
Add Qwen 3.6 Plus as third model in security review fleet via OpenRouter API. Refactor models-security-review.yml + new qwen-security-review.yml on top of a new shared reusable workflow `security-review-shared.yml`.

## Changes
- New workflow `.github/workflows/qwen-security-review.yml`
- New shared reusable workflow `.github/workflows/security-review-shared.yml`
- Refactor of `.github/workflows/models-security-review.yml` to call the shared workflow
- OpenRouter API endpoint with `OPENROUTER_API_KEY_QWEN` secret
- Updated `tools/check_models_security_review_workflow.py` contract checker to validate the shared workflow + both callers (`uses:` invocation, required inputs, Qwen API_KEY mapping)
- Model-specific anti-hallucination rules based on BridgeBench data:
  - 26.5% fabrication rate mitigation (FABRICATION PREVENTION)
  - Edge-case language behavior caution (async, closures, type coercion)
  - Security review limitations (43.3% hidden test success rate)

## Inherited from DeepSeek workflow
- Full parity context (60+ Go↔Rust pairs)
- Dependency contract extraction
- Previous findings deduplication (matched by full `model_display_name`)
- jsonrepair fallback
- `<think>` block stripping

## API Parameters (Qwen)
- Model: `qwen/qwen3.6-plus`
- temperature: 0.2
- top_p: 0.9
- max_tokens: 65000 (total budget)
- **reasoning.max_tokens: 50000** (caps the always-on chain-of-thought; ~15k tokens left for the structured JSON answer)
- response_format: json_object

DeepSeek caller uses the same `temperature: 0.2` and `top_p: 0.9` but does NOT send `reasoning` (it targets the GitHub Models chat endpoint, which is non-CoT and would reject the field).

## Testing
Requires `OPENROUTER_API_KEY_QWEN` secret to be configured in repository settings. Local validation:
- `python3 tools/check_models_security_review_workflow.py` → contract check (79 parity pairs)
- `python3 -m unittest tools.tests.test_check_models_security_review_workflow`

### CODE REVIEW

**Comment by 2tbmz9y2xt-lang** [comment]
Re: duplication concern from @cursor

**Intentional for MVP.** Each model security review workflow is self-contained:
- Different API endpoints (GitHub Models vs OpenRouter)
- Different auth (GITHUB_TOKEN vs custom secret)
- Model-specific anti-hallucination rules
- Independent failure modes

**Future:** Extract shared logic (parityMap, readChangedFile, acknowledged-list gate) into reusable workflow when adding 4th+ model. Current approach avoids cross-workflow breaking changes during rapid iteration.

File: .github/workflows/qwen-security-review.yml [hunk 1]
```diff
@@ -0,0 +1,786 @@
...
+                fs.writeFileSync('raw-qwen-output.txt', rawContent);
+                core.setFailed(`Security review fail-closed: jsonrepair also failed: ${e2.message}`);
+                return;
+              }
+            }
+
+            core.setOutput('review', result);
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          OPENROUTER_API_KEY_QWEN: ${{ secrets.OPENROUTER_API_KEY_QWEN }}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in 9cf1339: Added `check-key` step that skips workflow when `OPENROUTER_API_KEY_QWEN` unavailable (fork PRs, Dependabot).

File: .github/workflows/qwen-security-review.yml [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+                    messages: [
+                      { role: 'system', content: systemPrompt },
+                      { role: 'user', content: userContent }
+                    ],
+                    temperature: 0.2,
+                    top_p: 0.9,
+                    max_tokens: 65000,
+                    response_format: { type: 'json_object' }
+                  })
+                }
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in 9cf1339: `check-key` step emits `skip=true` when secret unavailable → `Run security review` step skipped via `if: steps.check-key.outputs.skip != 'true'`. No 401, no fail-close.

File: .github/workflows/qwen-security-review.yml [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+                    model: modelId,
+                    messages: [
+                      { role: 'system', content: systemPrompt },
+                      { role: 'user', content: userContent }
+                    ],
+                    temperature: 0.2,
+                    top_p: 0.9,
+                    max_tokens: 65000,
+                    response_format: { type: 'json_object' }
+                  })
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in 9cf1339: Added truncation before `createReview` — `maxReviewLen = 65000 - header.length - footer.length`, review JSON sliced to fit GitHub comment body limit.

File: .github/workflows/qwen-security-review.yml [hunk 4]
```diff
@@ -0,0 +1,786 @@
...
+            if (novel.length > 0) {
+              core.setFailed(`Merge blocked: ${novel.length} NEW CRITICAL/HIGH finding(s) (${suppressed.length} acknowledged, skipped)`);
+              for (const f of novel) {
+                core.error(`  [NEW] ${f.severity}: ${f.title} (${f.file})`);
+              }
+            } else {
+              core.info(`No new CRITICAL/HIGH findings — review passed (${suppressed.length} acknowledged finding(s) suppressed)`);
+            }
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Valid. Tracked as followup: extract shared logic (parityMap, readChangedFile, acknowledged-list gate) into reusable workflow after this PR merges.

File: .github/workflows/qwen-security-review.yml [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+        with:
+          fetch-depth: 0
+      - name: Download diff artifact
+        uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c  # v4
+        with:
+          name: qwen-pr-diff
+      - name: Check API key availability
+        id: check-key
+        run: |
+          if [ -z "${{ secrets.OPENROUTER_API_KEY_QWEN }}" ]; then
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in ec030cd: Secret passed via env, not direct interpolation.

File: .github/workflows/qwen-security-review.yml [hunk 6]
```diff
@@ -0,0 +1,803 @@
...
+            echo "skip=true" >> "$GITHUB_OUTPUT"
+            echo "::notice::OPENROUTER_API_KEY_QWEN not available (fork PR or missing secret). Skipping Qwen review."
+          else
+            echo "skip=false" >> "$GITHUB_OUTPUT"
+          fi
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in ec030cd: Secret passed via env.

File: .github/workflows/qwen-security-review.yml [hunk 7]
```diff
@@ -0,0 +1,803 @@
...
+                return;
+              }
+              rawContent = data.choices[0].message.content;
+              result = rawContent;
+            } catch (err) {
+              core.setFailed(`Security review fail-closed: model ${modelId} error: ${err.message}`);
+              return;
+            }
+
+            // Strip R1 <think> reasoning blocks
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in ec030cd: Comment updated to Qwen CoT.

File: .github/workflows/qwen-security-review.yml [hunk 8]
```diff
@@ -0,0 +1,803 @@
...
+                result = jsonrepair(result);
+                JSON.parse(result); // validate repair worked
+              } catch (e2) {
+                fs.writeFileSync('raw-qwen-output.txt', rawContent);
+                core.setFailed(`Security review fail-closed: jsonrepair also failed: ${e2.message}`);
+                return;
+              }
+            }
+
+            core.setOutput('review', result);
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml]: Fixed in ec030cd: Review written to file, filename passed via output. Avoids step output size limits.

**Comment by 2tbmz9y2xt-lang** [comment]
@codex review

File: .github/workflows/security-review-shared.yml [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+        default: false
+    secrets:
+      API_KEY:
+        description: "API key for the model provider"
+        required: true
+
+permissions:
+  contents: read
+  pull-requests: write
+
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml:56]: Fixed: added models:read to DeepSeek caller

File: .github/workflows/security-review-shared.yml [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+              core.setFailed(`Unparseable review JSON: ${e.message}`);
+              return;
+            }
+
+            // Load acknowledged findings from BASE ref
+            const baseSha = '${{ github.event.pull_request.base.sha }}';
+            let acknowledged = [];
+            try {
+              const content = execSync(
+                `git show ${baseSha}:.github/security-review-acknowledged.json 2>/dev/null || echo "[]"`,
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed: corrected path to .github/security-acknowledged.json and format parsing

File: .github/workflows/security-review-shared.yml [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+        type: boolean
+        default: false
+    secrets:
+      API_KEY:
+        description: "API key for the model provider"
+        required: true
+
+permissions:
+  contents: read
+  pull-requests: write
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml:49]: Fixed: models:read in DeepSeek caller

File: .github/workflows/security-review-shared.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed: correct path and format

File: .github/workflows/security-review-shared.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml:49]: Fixed: models:read added to DeepSeek caller permissions

File: .github/workflows/security-review-shared.yml [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+                parityEntries.push(entry);
+                parityUsed += separatorOverhead + introOverhead + entry.length;
+              }
+            }
+            const parityContext = parityEntries.length > 0 ? parityIntro + parityEntries.join(SEPARATOR) : '';
+
+            // Parse model-specific anti-hallucination rules
+            let antiHallucinationRules = [];
+            try {
+              antiHallucinationRules = JSON.parse('${{ inputs.anti_hallucination_rules }}');
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed: use toJSON() for safe interpolation

File: .github/workflows/security-review-shared.yml [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+            // Load acknowledged findings from BASE ref
+            const baseSha = '${{ github.event.pull_request.base.sha }}';
+            let acknowledged = [];
+            try {
+              const content = execSync(
+                `git show ${baseSha}:.github/security-review-acknowledged.json 2>/dev/null || echo "[]"`,
+                { encoding: 'utf8' }
+              );
+              acknowledged = JSON.parse(content);
+            } catch {}
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml:837]: Fixed: path corrected to .github/security-acknowledged.json

File: .github/workflows/security-review-shared.yml [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+
+            const isAcknowledged = (f) => acknowledged.some(ack => {
+              if (ack.file && ack.file !== f.file) return false;
+              if (ack.title_contains && !f.title?.includes(ack.title_contains)) return false;
+              if (ack.title_regex && !new RegExp(ack.title_regex).test(f.title || '')) return false;
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed: isAcknowledged now uses title_pattern and file_pattern

File: .github/workflows/models-security-review.yml [hunk 15]
```diff
@@ -4,761 +4,17 @@ on:
...
+    uses: ./.github/workflows/security-review-shared.yml
+    with:
+      model_id: "deepseek/DeepSeek-R1-0528"
+      api_base_url: "https://models.github.ai/inference/chat/completions"
+      model_display_name: "deepseek/DeepSeek-R1-0528"
+      artifact_prefix: "deepseek"
+      max_tokens: 0
+      needs_json_mode: false
+      uses_github_models: true
+      anti_hallucination_rules: "[]"
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/models-security-review.yml:23]: Fixed: pull-requests:write added to caller permissions

File: .github/workflows/qwen-security-review.yml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  qwen-review:
+    uses: ./.github/workflows/security-review-shared.yml
+    with:
+      model_id: "qwen/qwen3.6-plus"
+      api_base_url: "https://openrouter.ai/api/v1/chat/completions"
+      model_display_name: "qwen/qwen3.6-plus (OpenRouter)"
+      artifact_prefix: "qwen"
+      max_tokens: 65000
+      needs_json_mode: true
+      uses_github_models: false
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/qwen-security-review.yml:28]: Fixed: permissions added to Qwen caller

File: .github/workflows/security-review-shared.yml [hunk 17]
```diff
@@ -0,0 +1,667 @@
...
+                parityEntries.push(entry);
+                parityUsed += separatorOverhead + introOverhead + entry.length;
+              }
+            }
+            const parityContext = parityEntries.length > 0 ? parityIntro + parityEntries.join(SEPARATOR) : '';
+
+            // Parse model-specific anti-hallucination rules
+            let antiHallucinationRules = [];
+            try {
+              antiHallucinationRules = JSON.parse('${{ inputs.anti_hallucination_rules }}');
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Already fixed in c5f9398: using toJSON() for safe interpolation

File: .github/workflows/security-review-shared.yml [hunk 18]
```diff
@@ -0,0 +1,667 @@
...
+                  const sevOrder = ['CRITICAL','HIGH','MEDIUM','LOW','INFO','PERF'];
+                  if (sevOrder.indexOf(a.severity) < sevOrder.indexOf(b.severity)) dominated.add(j);
+                }
+              }
+            }
+
+            const dominated_arr = [...dominated];
+            const dominated_filtered = findings.filter((_, i) => !dominated_arr.includes(i));
+            const dominated_unacked = dominated_filtered.filter(f => !isAcknowledged(f));
+            const dominated_unacked_blocking = dominated_unacked.filter(f => f.severity === 'CRITICAL' || f.severity === 'HIGH');
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed in 456a44f: added normSev() to uppercase/trim severity before comparison

File: .github/workflows/security-review-shared.yml [hunk 19]
```diff
@@ -0,0 +1,667 @@
...
+              '```',
+              previousFindings
+            ].join('\n');
+
+            const userPrompt = [
+              'Review this PR diff for security issues:\n',
+              diff,
+              changedFilePayload,
+              dependencyContext,
+              parityContext
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed: userPrompt.slice(0, 300000) added

File: .github/workflows/security-review-shared.yml [hunk 20]
```diff
@@ -0,0 +1,667 @@
...
+
+            const dominated = new Set();
+            const findings = parsed.findings || [];
+            for (let i = 0; i < findings.length; i++) {
+              for (let j = 0; j < findings.length; j++) {
+                if (i === j) continue;
+                const a = findings[i], b = findings[j];
+                if (a.file === b.file && a.title === b.title) {
+                  const sevOrder = ['CRITICAL','HIGH','MEDIUM','LOW','INFO','PERF'];
+                  if (sevOrder.indexOf(a.severity) < sevOrder.indexOf(b.severity)) dominated.add(j);
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml]: Fixed: getSevIdx() returns sevOrder.length for unknown severities (lowest priority)

File: .github/workflows/security-review-shared.yml [hunk 21]
```diff
@@ -0,0 +1,671 @@
...
+                }
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml:899]: Fixed in 08e00e6: getSevIdx() returns sevOrder.length for unknown severities

File: .github/workflows/security-review-shared.yml [hunk 22]
```diff
@@ -0,0 +1,671 @@
...
+              }
+            }
+
+            const dominated_arr = [...dominated];
+            const dominated_filtered = findings.filter((_, i) => !dominated_arr.includes(i));
+            const dominated_unacked = dominated_filtered.filter(f => !isAcknowledged(f));
+            const normSev = (s) => String(s || '').toUpperCase().trim();
+            const dominated_unacked_blocking = dominated_unacked.filter(f => {
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/security-review-shared.yml:909]: Fixed in 456a44f: normSev() normalizes to uppercase before comparison

File: .github/workflows/models-security-review.yml [hunk 23]
```diff
@@ -10,755 +10,15 @@ permissions:
...
+    secrets: inherit
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply .github/workflows/models-security-review.yml]: Fixed: updated contract check script for refactored structure - now checks security-review-shared.yml

**Comment by 2tbmz9y2xt-lang** [comment]
@codex review
